PLIC Cleanup

FawazTirmizi opened the pull request:
Fixes a bug where Idol functions that the server returned an error for were not returning errors to the client. Also fixes PLIC interrupt behavior to be in line with what is expected of interrupts delivered by the kernel at startup, and changed `mie::mext` to be disabled at startup.

FawazTirmizi committed:
Fix bug where ext_int_ctrl idol calls never failed

This bug was caused by the only error code,
`ExtIntCtrlError::IRQUnassigned` being equal to zero. Idol clients
interpret only nonzero return values as errors, so changing the enum to
1 fixes this bug.

furquan-rivos commented on drv/ext-int-ctrl-api/src/lib.rs:
```diff
@@ -9,7 +9,7 @@ use userlib::*;
 #[derive(Copy, Clone, Debug, FromPrimitive, IdolError)]
 pub enum ExtIntCtrlError {
-    IRQUnassigned,
+    IRQUnassigned = 1,
```
Is there a way to add a compile time check to ensure the error codes are always non-zero?

FawazTirmizi replied:
Not really, Rust unfortunately doesn't have any form of static assert yet. We could add something to the `build.rs` that would check this, but that would still require new idol interfaces to include that `build.rs`. Maybe there is some way to add this check to [idolatry](https://github.com/oxidecomputer/idolatry), but I'm not totally sure of how we'd do it.

beezow replied:
We could also use something like [this](https://docs.rs/static_assertions/0.1.1/static_assertions/index.html).  It seems to be the standard way of doing compile time checks right now.

furquan-rivos replied:
Ack. I think we can follow this up on a separate JIRA issue so that we don't block this PR.

beezow replied:
Made a [ticket](https://rivosinc.atlassian.net/browse/SW-599)

FawazTirmizi committed:
Leave PLIC interrupts disabled by default

Originally, the PLIC would unmask interrupts at startup. This is not
what the kernel does for core interrupts,, so this behavior has been
changed.

FawazTirmizi committed:
Leave `mie::mext` disabled on startup

Because external interrupts are now delegated to tasks, enabling them in
mie must also be delegated to the task handling them.

beezow commented on drv/riscv-plic-server/src/main.rs:
```diff
@@ -174,7 +174,6 @@ fn main() -> ! {
     // Zero all interrupt sources ...
```
Out of date comment.  Should probably update to reflect the fact you are intentionally not zeroing the plic.

FawazTirmizi replied:
Done
